### PR TITLE
Implement customcli

### LINKED ISSUE #5: Build sample cli
Demonstrate how to connect to an app.

Use `cmd/bnscli` from weave as a basis. Ideally we can refactor that to make code reuse possible, and then just a few lines to show a command line interface to send transactions and query balances.

This will allow the full app demo in one golang repo.

### PR DESCRIPTION
**Pull request by orkunkl**
This Pull Request implements `customcli` (cleaned  up bnscli) 
Resolves https://github.com/iov-one/weave-starter-kit/issues/5

### CODE REVIEW

**Comment by husio** [comment]
I wonder if it makes sense to copy `bnscli` to a starter kit. `bnscli` is not an officially supported tool. Another reason is that this code is not portable to new applications. You have to copy paste it and modify in order to work with your application. This makes it easy to introduce bugs.

What is the purpose of `customcli`? Is is an example of how to write a console client or this is supposed to be a base for writing a custom application client?

**Comment by orkunkl** [comment]
We should provide an easy to customize client codebase that users can base their work upon. Thus copy-paste and modify process is not mandatory, one can [fork](https://dev.to/loderunner/working-with-forks-in-go-3ab6) and then [replace directive](https://github.com/golang/go/wiki/Modules#when-should-i-use-the-replace-directive). This way users can get updates from `weave-starter-kit`. Also whenever they feel like their project is ready to be unlinked from `weave-starter-kit`, they can easily do the change.

**Comment by orkunkl** [comment]
Thus this task https://github.com/iov-one/weave-starter-kit/issues/5 has been given me by Ethan

**Comment by husio** [comment]
I think those are valid point and it is a great functionality to provide to a client. I wonder if the same could be achieved with less effort required from the user. `bnscli` was created without portability in mind. It is implementing functionality that was needed without fully supporting each extension. If we ever update or add an extension, you must know which code to copy to your client. I think this might be an issue.


If we could make each command function (for example `cmdSendTokens`, `cmdMultisig`) a public function of each extension that is imporoted by the CLI tool, we would provide a better support for writing command line tools. As a CLI author, you only have to import them and install in your program. You are getting all updates together with weave updates. You do not have to test those functions because they are coming from weave.

What do you think?

**Comment by husio** [comment]
I have tried to port a single command to be implemented by the extension. I have picked the `x/escrow` and there is a lot of work required in order to make this possible. The problem is that `bnscli` imports `bnsd.Tx` structure and directly access it. This is [not possible if we want to keep the extension code generic](https://github.com/iov-one/weave/commit/935c875e1df0b647f1ebb4b06e73b5767fb97513#diff-910abec62c1d78d9eebcb520c2ef0f63R34-R35). Instead, we would have to figure out how to provide the functionality not knowing what is the transaction implementation.

File: cmd/customcli/cmd_query.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+			return fmt.Errorf("can not encode data: %s", err)
+		}
+	}
+	queryPath := *pathFl
+	if *prefixQueryFl || *dataFl == "" {
+		queryPath += "?" + weave.PrefixQueryMod
+	}
+
+	bnsClient := client.NewClient(client.NewHTTPConnection(*tmAddrFl))
+	resp, err := bnsClient.AbciQuery(queryPath, data)
```
Comment (husio) [inline cmd/customcli/cmd_query.go]: No longer a bns client.

File: cmd/customcli/Makefile [hunk 2]
```diff
@@ -1 +1,15 @@
+all: install
+
+build:
+	CGO_ENABLED=0 go build -mod=readonly -ldflags "-X main.gitHash=$(shell git rev-list -1 HEAD) -extldflags \"-static\"" .
+
+
+clean:
+	-rm bnscli
```
Comment (husio) [inline cmd/customcli/Makefile]: No longer a `bnscli`

File: cmd/customcli/main.go [hunk 3]
```diff
@@ -1,5 +1,77 @@
...
+// a pipeline. For example, there are 3 separate functions for creating a
+// transaction, signing and submitting. They can be combined into a single
+// pipeline line:
+//
+//   $ customcli release-escrow -escrow 1 \
+//       | customcli as-proposal \
+//       | customcli sign \
+//       | customcli submit
+//
+var commands = map[string]func(input io.Reader, output io.Writer, args []string) error{}
```
Comment (husio) [inline cmd/customcli/main.go]: empty `commands` might be confusing. Maybe add a comment as a body explaining how to register functions.

**Review by husio — APPROVED** [review]

**Pull request merged**